src/service/video-manager-impl.js

```diff
@@ -0,0 +1,310 @@
+
+    /** @const @private {!../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(win);
+
+    /** @private {boolean} */
+    this.canAutoplay_ = video.hasAutoplay() &&
+        platformSupportsAutoplay(platformFor(win));
+
+    const element = dev().assert(video.element);
+
+    listenOncePromise(element, VideoEvents.BUILT)
+      .then(() => this.videoBuilt_());
+
+    listenOncePromise(element, VideoEvents.CAN_PLAY)
+      .then(() => this.videoLoaded_());
+  }
+
+  /**
+   * Called when the video element is built.
+   * @private
+   */
+  videoBuilt_() {
+    if (this.canAutoplay_) {
+      this.autoplayVideoBuilt_();
+    }
+  }
+
+  /**
+   * Called when the video is loaded and can play.
+   * @private
+   */
+  videoLoaded_() {
+    this.loaded_ = true;
+    if (this.isVisible_) {
+      // Handles the case when the video becomes visible before loading
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Called when visibility of a video changes.
+   * @private
+   */
+  videoVisibilityChanged_() {
+    if (this.loaded_) {
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Only called when visibility of a loaded video changes.
+   * @private
+   */
+  loadedVideoVisibilityChanged_() {
+    if (this.canAutoplay_) {
+      this.autoplayLoadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Whether there is a need to monitor visibility of this video.
+   * @return {boolean}
+   * @package
+   */
+  needsVisibilityObserver() {
+    return this.canAutoplay_;
+  }
+
+  /* Autoplay Behaviour */
+
+  /**
+   * Called when an autoplay video is built.
+   * @private
+   */
+  autoplayVideoBuilt_() {
+    this.video.mute();
+
+    // If autoplay video has controls, hide them and only show them on
+    // user-ineraction.
+    if (this.video.hasControls()) {
+      this.video.hideControls();
+
+      // TODO(aghassemi): This won't work for iframes, needs a transparent shim
+      listenOnce(this.video.element, 'click', () => {
+        this.video.showControls();
+        this.video.unmute();
+      });
+    }
+  }
+
+  /**
+   * Called when visibility of a loaded autoplay video changes.
+   * @private
+   */
+  autoplayLoadedVideoVisibilityChanged_() {
+    if (this.isVisible_) {
+      this.video.play(/*autoplay*/ true);
+    } else {
+      this.video.pause();
+    }
+  }
+
+  /**
+   * Called by all possible events that might change the visibility of the video
+   * such as scrolling or {@link ../video-interface.VideoEvents#VISIBILITY}.
+   * @package
+   */
+  updateVisibility() {
+    const wasVisible = this.isVisible_;
+
+    // Measure if video is now in viewport and what percentage of it is visible
+    const measure = () => {
+      if (!this.video.isInViewport()) {
+        this.isVisible_ = false;
+        return;
+      }
+
+      // Calculate what percentage of the video is in viewport
+      const change = this.video.element.getIntersectionChangeEntry();
```

Comment:
This should be documented properly, but I believe it is fine to call this without a `measure` block, because it exclusively hits methods that are cached and only written to when things change.